frameworks.qmd

```diff
@@ -1,119 +1,1793 @@
+optimization algorithms, etc.). Examples of high-level frameworks
+include TensorFlow/Keras and PyTorch. Examples of low-level ML
+frameworks include TensorFlow with low-level APIs, Theano, Caffe,
+Chainer, and CNTK.
+
+Frameworks like Theano and Caffe used static computational graphs which
+required rigidly defining the full model architecture upfront. Static
+graphs require upfront declaration and limit flexibility. Dynamic graphs
+construct on-the-fly for more iterative development. But around 2016,
+frameworks began adopting dynamic graphs like PyTorch and TensorFlow 2.0
+which can construct graphs on-the-fly. This provides greater flexibility
+for model development. We will discuss these concepts and details later
+on in Section XXX.
+
+The development of these frameworks facilitated an explosion in model
+size and complexity over time---from early multilayer perceptrons and
+convolutional networks to modern transformers with billions or trillions
+of parameters. In 2017, ResNet models achieved record ImageNet accuracy
+with over 150 layers and 25 million parameters. Then in 2020, the GPT-3
+language model pushed parameters to an astonishing 175 billion using
+model parallelism in frameworks to train across thousands of GPUs and
+TPUs.
+
+Each generation of frameworks unlocked new capabilities that powered
+advancement:
+
+-   Theano and TensorFlow (2015) introduced computational graphs and
+    > automatic differentiation to simplify model building.
+
+-   CNTK (2016) pioneered efficient distributed training by combining
+    > model and data parallelism.
+
+-   PyTorch (2016) provided imperative programming and dynamic graphs
+    > for flexible experimentation.
+
+-   TensorFlow 2.0 (2019) made eager execution default for intuitiveness
+    > and debugging.
+
+-   TensorFlow Graphics (2020) added 3D data structures to handle point
+    > clouds and meshes.
+
+In recent years, there has been a convergence on the frameworks.
+TensorFlow and PyTorch have become the overwhelmingly dominant ML
+frameworks, representing more than 95% of ML frameworks used in research
+and production. Keras was integrated into TensorFlow in 2019; Preferred
+Networks transitioned Chainer to PyTorch in 2019; and Microsoft stopped
+actively developing CNTK in 2022 in favor of supporting PyTorch on
+Windows.
+
+![](images_ml_frameworks/image6.png){width="3.821385608048994in"
+height="2.5558081802274715in"}
+
+Popularity of ML frameworks in the United States as measured by Google
+web searches
+
+However, a one-size-fits-all approach does not work well across the
+spectrum from cloud to tiny edge devices. Different frameworks represent
+various philosophies around graph execution, declarative versus
+imperative APIs, and more. Declarative defines what the program should
+do while imperative focuses on how it should do it step-by-step. For
+instance, TensorFlow uses graph execution and declarative-style modeling
```

Comment:
It would be helpful to provide more explanation on the difference between declarative and imperative APIs. Consider giving an example of the actual function of a declarative API vs an imperative API instead of just citing TensorFlow and Pytorch.